content/academy/pro_apify_scraper_development/actor_building.md

```diff
@@ -0,0 +1,518 @@
+
+```JavaScript
+exports.handleProduct = ({ $, crawler: { requestQueue }, request }) => {
+    const { data } = request.userData;
+
+    const element = $('div#productDescription');
+
+    const description = element.text().trim();
+
+    console.log(description); // works!
+};
+```
+
+> If you are sometimes getting an error along the lines of **RequestError: Proxy responded with 407**, don't worry, this is totally normal. The proxy being used for the request was probably dead, so the request failed, but CheerioCrawler will automatically retry the request.
+
+Great! But wait, where do we go from here? We need to go to the offers page next and scrape each offer, but how can we do that? Let's take a small break from writing the scraper and open up [Proxyman]({{@link glossary/proxyman.md}}) to analyze requests which we can't see inside the network tab, then we'll click the button on the product page that loads up all of the product offers:
+
+![View offers button]({{@asset pro_apify_scraper_development/images/view-offers-button.webp}})
+
+After clicking this button and checking back in Proxyman, we discovered this link:
+
+```text
+https://www.amazon.com/gp/aod/ajax/ref=auto_load_aod?asin=B07ZPKBL9V&pc=dp
+```
+
+The `asin` parameter matches up with our product's ASIN, which means we can use this for any product of which we have the ASIN.
+
+Here's what this page looks like:
+
+![View offers page]({{@asset pro_apify_scraper_development/images/offers-page.webp}})
+
+Wow, that's ugly. But for our scenario, this is really great. When we click the **View offers** button, we usually have to wait for the offers to load and render, which would mean we could have to switch our entire crawler to a **PuppeteerCrawler** or **PlaywrightCrawler**. The data on this page we've just found appears to be loaded statically, which means we can still use CheerioCrawler and keep the scraper as efficient as possible 😎
+
+First, we'll create a function which can generate an offers URL for us in **constants.js**:
+
+```JavaScript
+const OFFERS_URL = (asin) => `${BASE_URL}/gp/aod/ajax/ref=auto_load_aod?asin=${asin}&pc=dp`;
+```
+
+Then, we'll import and use that function to create a request for each product's offers page:
+
+```JavaScript
+exports.handleProduct = ({ $, crawler: { requestQueue }, request }) => {
+    const { data } = request.userData;
+
+    const element = $('div#productDescription');
+
+    // Return our promise
+    return requestQueue.addRequest({
+        url: OFFERS_URL(data.asin),
+        userData: {
+            label: labels.OFFERS,
+            data: {
+                ...data,
+                description: element.text().trim(),
+            },
+        },
+    });
+};
+```
+
+## [](#final-code) Final code
+
+That should be it! Let's just make sure we've all got the same code:
+
+```JavaScript
+// src/constants.js
```

Comment:
For these, could you put the file name in text above the code block, please? Maybe even as 3rd-level headings, if you think that's suitable (or maybe just bold?)?
```suggestion

**src/constants.js**

```JavaScript
// src/constants.js
```